Add grpc_socket_mutator

y-zeng opened the pull request:
- Add grpc_socket_mutator,  which can be used to set the socket options of a channel.
- Support grpc_socket_mutator in ChannelArguments and grpc_channel_args.
- ~Introduce grpc_tcp_client_connect_args, which helps pass channel arguments to the tcp client~

To set the socket options of a channel:
1. Implement a grpc_socket_mutator
2. Create an instance of the implemented grpc_socket_mutator
3. Add this instance to a ChannelArguments instance using ChannelArguments::SetChannelArgs
4. Create the channel with this ChannelArguments instance using CreateCustomChannel

ctiller commented on test/cpp/common/channel_arguments_test.cc:
```diff
@@ -35,10 +35,50 @@
+  delete tsm;
+}
+
+grpc_socket_mutator_vtable test_mutator_vtable = {test_mutator_mutate_fd,
```
Please implement a cmp function (could just compare the 'this' pointers): we should always fill out all of the vtable whenever we can.

y-zeng replied:
Thanks a lot for the review!
The cmp function in `grpc_arg_pointer_vtable` is implemented in [socket_mutator.c](https://github.com/grpc/grpc/pull/8335/files#diff-66f165bfbf516848b03e99c9bedd5254R69). It compares the pointers using `GPR_ICMP()`.
`grpc_socket_mutator_vtable` does not have a cmp function. Should we add one?

y-zeng commented:
As discussed offline, added `compare()` in `grpc_socket_mutator_vtable` and added `grpc_socket_mutator_compare()`.
When comparing two socket mutator channel arguments, the pointer addresses are compared first. If they are not the same instance but have the same vtable, they will be compared using the `compare()` method in the vtable.

y-zeng commented:
Test failures seem unrelated. Only `grpc_channel_args_test`, `socket_utils_test` and `channel_arguments_test` should be affected.
